calyx/src/lang/component.rs

```diff
@@ -119,84 +119,84 @@ impl ast::ComponentDef {
     //        Structure Helper Functions
     //==========================================
 
+    fn from_structure<F, T>(&self, to_data: F) -> Vec<&T>
+    where
+        F: FnMut(&ast::Structure) -> Option<&T>,
+    {
+        self.structure.iter().filter_map(to_data).collect()
+    }
+
+    /// Get all the wires in this Component.
     pub fn get_wires(&self) -> Vec<&ast::Wire> {
-        let mut v: Vec<&ast::Wire> = Vec::new();
-        for structure in self.structure.iter() {
-            if let ast::Structure::Wire { data } = structure {
-                v.push(data)
-            }
-        }
-        v
+        self.from_structure(|s| match s {
+            ast::Structure::Wire { data } => Some(data),
+            _ => None,
+        })
     }
 
+    /// Get all `Std` from this Component.
     pub fn get_std(&self) -> Vec<&ast::Std> {
-        let mut v: Vec<&ast::Std> = Vec::new();
-        for structure in self.structure.iter() {
-            if let ast::Structure::Std { data } = structure {
-                v.push(data)
-            }
-        }
-        v
+        self.from_structure(|s| match s {
+            ast::Structure::Std { data } => Some(data),
+            _ => None,
+        })
     }
 
+    /// Get all `Decl` from this Component.
     pub fn get_decl(&self) -> Vec<&ast::Decl> {
-        let mut v: Vec<&ast::Decl> = Vec::new();
-        for structure in self.structure.iter() {
-            if let ast::Structure::Decl { data } = structure {
-                v.push(data)
-            }
-        }
-        v
+        self.from_structure(|s| match s {
+            ast::Structure::Decl { data } => Some(data),
+            _ => None,
+        })
     }
 
+    /// Build a `CompStore` that contains a mapping from component and Decl
+    /// names to clones of the componenets.
     pub fn get_store(&self) -> CompStore {
-        let mut store: CompStore = HashMap::new();
-        let std = self.get_std();
-        let new = self.get_decl();
-        for inst in std {
-            store.insert(
-                inst.name.clone(),
-                ast::Structure::Std { data: inst.clone() },
```

Comment:
Why does this return `clones` of the structure graph components instead of references? What is the use case for this function?

Reply:
I'm transitioning code away from using CompStore. The Context replaces this. We can honestly remove most of these functions and switch them over to acting on the context directly.